File: aws-mfa-auth.py
import argparse
import os
import boto3

def get_temp_credentials(profile_name, token_code, mfa_serial):
    # Create a session with the specified profile
    session = boto3.Session(profile_name=profile_name)

    # Get temporary credentials using the STS client and MFA token code
    client = session.client('sts')
    response = client.get_session_token(
        DurationSeconds=3600,
        SerialNumber=mfa_serial,
        TokenCode=token_code
    )

    # Return the temporary credentials
    return response['Credentials']

if __name__ == '__main__':
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Get temporary AWS credentials using MFA')
    parser.add_argument('--profile', required=True, help='Name of the AWS CLI profile to use')
    parser.add_argument('--token', required=True, help='MFA token code')
    parser.add_argument('--mfa-serial', required=True, help='ARN of the MFA device')
    args = parser.parse_args()

    # Get temporary credentials
    temp_creds = get_temp_credentials(args.profile, args.token, args.mfa_serial)

    # Set environment variables with the temporary credentials
    os.environ['AWS_ACCESS_KEY_ID'] = temp_creds['AccessKeyId']
    os.environ['AWS_SECRET_ACCESS_KEY'] = temp_creds['SecretAccessKey']
    os.environ['AWS_SESSION_TOKEN'] = temp_creds['SessionToken']

    # Print a message indicating that the environment variables have been set
    print("Temporary credentials obtained with MFA. Environment variables have been set.")

    # Print the temporary credentials
    print(f"AWS_ACCESS_KEY_ID={temp_creds['AccessKeyId']}")
    print(f"AWS_SECRET_ACCESS_KEY={temp_creds['SecretAccessKey']}")
    print(f"AWS_SESSION_TOKEN={temp_creds['SessionToken']}")

File: aws-mfa-auth.go
package main

import (
	"flag"
	"fmt"
	"os"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/sts"
)

func getTempCredentials(profileName string, tokenCode string, mfaSerial string) (*sts.Credentials, error) {
	// Create a session with the specified profile
	sess, err := session.NewSessionWithOptions(session.Options{
		Profile: profileName,
	})

	if err != nil {
		return nil, err
	}

	// Get an STS client using the session
	svc := sts.New(sess)

	// Get temporary credentials using the STS client and MFA token code
	params := &sts.GetSessionTokenInput{
		DurationSeconds: aws.Int64(3600),
		SerialNumber:    aws.String(mfaSerial),
		TokenCode:       aws.String(tokenCode),
	}
	resp, err := svc.GetSessionToken(params)

	if err != nil {
		return nil, err
	}

	// Return the temporary credentials
	return resp.Credentials, nil
}

func main() {
	// Parse command line arguments
	profileName := flag.String("profile", "", "Name of the AWS CLI profile to use")
	tokenCode := flag.String("token", "", "MFA token code")
	mfaSerial := flag.String("mfa-serial", "", "ARN of the MFA device")
	flag.Parse()

	// Get temporary credentials
	tempCreds, err := getTempCredentials(*profileName, *tokenCode, *mfaSerial)

	if err != nil {
		fmt.Println(err)
		return
	}

	// Set environment variables with the temporary credentials
	os.Setenv("AWS_ACCESS_KEY_ID", *tempCreds.AccessKeyId)
	os.Setenv("AWS_SECRET_ACCESS_KEY", *tempCreds.SecretAccessKey)
	os.Setenv("AWS_SESSION_TOKEN", *tempCreds.SessionToken)

	// Print a message indicating that the environment variables have been set
	fmt.Println("Temporary credentials obtained with MFA. Environment variables have been set.")

	// Print the temporary credentials
	fmt.Printf("AWS_ACCESS_KEY_ID=%s\n", *tempCreds.AccessKeyId)
	fmt.Printf("AWS_SECRET_ACCESS_KEY=%s\n", *tempCreds.SecretAccessKey)
	fmt.Printf("AWS_SESSION_TOKEN=%s\n", *tempCreds.SessionToken)
}
